BLE porting update (#1359)
ARMmbed/Handbook: docs/porting/connectivity/ble.md

# Bluetooth Low Energy (BLE)

> Mbed OS uses the Cordio stack for its BLE host. We recommend porting the Cordio platform<!--is that what they call it?--> to your board, or porting at the HCI level<!--what's the distinction between porting the platform and porting the HCI level? If one is ported to the board, where is the other ported to? The chip? What's the difference between porting to a whole board v porting to a chip?--> as described in the HCI documentation.

Reviewer: API level - you port the whole stack, and provide the same interface mbed ble provides.
HCI level - the stack is split into two parts Host and Controller, HCI stands for host controller interface, the communication across the HCI is well defined and as long as you handle all the messages you don't care who the host is and viceversa, if you have a chip with its own controller then that's the level you want to port at (like wb55)
Cordio level - you use cordio for everything and you port the platform layer of cordio so that cordio can run on your board directly like the nrf52

Author: So the API level port is what we describe doing on this page, and the Cordio and HCI level ports are what the [Cordio docs](https://os.mbed.com/docs/mbed-cordio/19.02/introduction/index.html) describe?

Reviewer: Yes. I don't expect API porting to be in high demand, it's just here as an option since there's no reason to stop people from doing it but I expect ports to use Cordio.

> Mbed OS uses the Cordio stack for its BLE host. We recommend porting the Cordio platform<!--is that what they call it?--> to your board, or porting at the HCI level<!--what's the distinction between porting the platform and porting the HCI level? If one is ported to the board, where is the other ported to? The chip? What's the difference between porting to a whole board v porting to a chip?--> as described in the HCI documentation.

Author: ```suggestion
Mbed OS uses the Cordio stack for its BLE host. We recommend porting either the full Cordio stack, or the Host Controller Interface (HCI), as explained in the Cordio documentation. If you want to port your own BLE stack to replace the Cordio stack, follow the instructions on this page. 

 ```

<!--we need a decision from Andy about the Cordio docs before we start linking. Ours are 18 months old!-->

## Replacing the Cordio stack

> You can replace the Cordio stack by porting a complete alternative stack. You must do this at the API level<!--as opposed to which level?--> by:

Reviewer: As opposed to porting Cordio or porting at HCI level.

> You can replace the Cordio stack by porting a complete alternative stack. You must do this at the API level<!--as opposed to which level?--> by:

Author: ```suggestion
You can replace the Cordio stack by porting a complete alternative stack. You must port the full API by:
```

Reviewer: that is somewhat misleading, if you're porting the API then HCI is completely irrelevant and down to your implementation.

Author: Is this one okay now?

1. Disabling the Cordio compilation.
1. Implementing all the user interfaces (Gap, GattServer, GattClient, SecurityManager).
1. Providing a BLEInstaceBase that `createBLEInstance()` returns.

### Disabling Cordio

All the source files for Cordio are part of a library and are currently required by the BLE feature. Remove this requirement so that none of the Cordio implementations are compiled, otherwise they will conflict with your implementation.

> ### Implementing API classes and BLEInstanceBase

Author: ```suggestion
### Implementing the user interface APIs and BLEInstanceBase
```

Reviewer: BLEInstanceBase is not a user interface, it's the implementation of BLE used by the BLE class which is the user interface. BLEInstanceBase is the internal interface that allows the BLE class use your private implementation of the BLE stack.

Reviewer: It's a porting interface, not user by users.

Author: But it also has its own header, doesn't it? Maybe we should move the bit about BLEIstanceBase to the next section?

Reviewer: Well, they have different function (one is the user facing API, the other is an internal porting interface) but the porting process of implementing them is the same. Originally I had this as separate sections but they were identical (save for names) so I combined them

There are four user API files in`include/ble/`:

- Gap.h
- GattClient.h
- GattServer.h
- SecurityManager.h

And `BLEInstanceBase` is in `ble/internal/BLEInstanceBase.h`.

Each API contains a class in a namespace `ble::interface`. Your implementation must inherit from this class and have the same class name, but be directly in the `ble` interface. Your implementation classes must implement all interface functions.

Place your `include`<!--adding the bunny ears just helps parse the sentence--> files in a path ending with `ble/internal/`, and add the suffix `Impl`: `GapImpl.h`, `GattClientImpl.h`, `GattServerImpl.h`, `SecurityManagerImpl.h`, `BLEInstanceBaseImpl.h`.

### Providing BLEInstanceBase

The global function `createBLEInstance()` must provide your implementation of `BLEInstanceBase`.

> `BLEInstanceBase` must instantiate all the user interfaces and provide BLE event handling (see the BLEInstanceBase documentation for details<!--where is that?-->). `BLEInstanceBase` can also instantiate the PAL interfaces provided in `ble/internal`, depending on your implementation. They<!--who is they?--> may be used as the user interfaces, but are not required for user API level porting.

Reviewer: I mean the BLEInstanceBase comments in the header.

Author: This one? [https://os.mbed.com/docs/mbed-os/v6.2/mbed-os-api-doxy/_b_l_e_instance_base_8h.html](https://os.mbed.com/docs/mbed-os/v6.2/mbed-os-api-doxy/_b_l_e_instance_base_8h.html)

Seems a bit empty - this is the source: [https://os.mbed.com/docs/mbed-os/v6.2/mbed-os-api-doxy/_b_l_e_instance_base_8h_source.html](https://os.mbed.com/docs/mbed-os/v6.2/mbed-os-api-doxy/_b_l_e_instance_base_8h_source.html)

Or the class [https://os.mbed.com/docs/mbed-os/v6.2/mbed-os-api-doxy/class_b_l_e_instance_base.html](https://os.mbed.com/docs/mbed-os/v6.2/mbed-os-api-doxy/class_b_l_e_instance_base.html)

Reviewer: Don't know which html file best link to but as you can see the source shows the header has detailed instructions.

Author: ```suggestion
`BLEInstanceBase` must instantiate all the user interfaces and provide BLE event handling (see the [BLEInstanceBase documentation for details](../mbed-os-api-doxy/_b_l_e_instance_base_8h_source.html)). `BLEInstanceBase` can also instantiate the PAL interfaces provided in `ble/internal`, depending on your implementation. You can use the PAL classes by inheriting from the interface classes (like the user interfaces above), but they are not required for user API level porting.
```

> `BLEInstanceBase` must instantiate all the user interfaces and provide BLE event handling (see the BLEInstanceBase documentation for details<!--where is that?-->). `BLEInstanceBase` can also instantiate the PAL interfaces provided in `ble/internal`, depending on your implementation. They<!--who is they?--> may be used as the user interfaces, but are not required for user API level porting.

Reviewer: ```suggestion
`BLEInstanceBase` must instantiate all the user interfaces and provide BLE event handling (see the BLEInstanceBase documentation for details<!--where is that?-->). `BLEInstanceBase` can also instantiate the PAL interfaces provided in `ble/internal`, depending on your implementation. The PAL layer classes may be used by inheriting from interfaces classes in a similar manner as the user interfaces, but are not required for user API level porting.
```
